src/Microsoft.OData.Core/UriParser/Visitors/SelectPropertyVisitor.cs

```diff
@@ -151,17 +156,38 @@ private void ProcessTokenAsPath(NonSystemToken tokenIn)
                         break;
                     }
 
-                    // This means last segment a collection of complex type,
-                    // current segment can only be type cast and cannot be property name.
-                    if (currentLevelType == null)
+                    if (primitiveType == null && dynamicPath == null)
                     {
-                        currentLevelType = collectionType.ElementType.Definition as IEdmStructuredType;
-                    }
+                        // This means last segment a collection of complex type,
+                        // current segment can only be type cast and cannot be property name.
+                        if (currentLevelType == null)
+                        {
+                            currentLevelType = collectionType.ElementType.Definition as IEdmStructuredType;
+                        }
 
-                    // If there is no collection type in the path yet, will try to bind property for the next token
-                    // first try bind the segment as property.
-                    lastSegment = SelectPathSegmentTokenBinder.ConvertNonTypeTokenToSegment(nextToken, this.model,
-                        currentLevelType, resolver);
+                        // If there is no collection type in the path yet, will try to bind property for the next token
+                        // first try bind the segment as property.
+                        lastSegment = SelectPathSegmentTokenBinder.ConvertNonTypeTokenToSegment(nextToken, this.model,
+                            currentLevelType, resolver);
+                    }
+                    else
+                    {
+                        // determine whether we are looking at a type cast or a dynamic path segment.
+                        EdmPrimitiveTypeKind nextTypeKind = EdmCoreModel.Instance.GetPrimitiveTypeKind(nextToken.Identifier);
+                        IEdmPrimitiveType castType = EdmCoreModel.Instance.GetPrimitiveType(nextTypeKind);
+                        if (castType != null)
+                        {
+                            lastSegment = new TypeSegment(castType, castType, null);
+                        }
+                        else if (dynamicPath != null)
+                        {
+                            lastSegment = new DynamicPathSegment(nextToken.Identifier);
+                        }

```

Comment:
the difference is just sugar and would force the type derivation to occur even in cases when it is not necessary.

---
In reply to: [124894287](https://github.com/OData/odata.net/pull/848#discussion_r124894287) [](ancestors = 124894287)